### PR TITLE
Add user trust attributes plugin

### PR DESCRIPTION
**Pull request by Tiboris**
Add user trust attributes plugin
https://github.com/abbra/freeipa-user-trust-attributes
Author: @abbra Alexander Bokovoy <abokovoy@redhat.com>

### CODE REVIEW

**Comment by abbra** [comment]
@Tiboris could you please merge this code into `ipaserver/plugins/user.py` instead of having it as a separate file? It only was in a separate file because the `user-trust-attributes` was an external plugin.

**Comment by abbra** [comment]
Same with UI changes -- this code is amending existing user.js definition but instead should be merged into that one.

**Comment by Tiboris** [comment]
Ok, got it.

File: ipaserver/install/adtrustinstance.py [hunk 1]
```diff
@@ -482,6 +482,40 @@ def __add_cldap_module(self):
         """
         self.__add_plugin_conf('CLDAP', 'ipa_cldap', 'ipa-cldap-conf.ldif')
 
+    def __allow_trust_attr_mod(self):
+        """
+        aci to allow users to manage own trust attributes
+        """
+        # user_aci = ipalib.aci.ACI(
+        user_aci = \
```
Comment (tiran) [inline ipaserver/install/adtrustinstance.py]: This function looks unnecessarily complicated. You should either be able to create the permission from ``user.managed_permissions`` dict or from a regular update file in ``./install/updates/``.

Comment (Tiboris) [reply ipaserver/install/adtrustinstance.py]: Thanks @tiran.

File: ipaserver/install/adtrustinstance.py [hunk 2]
```diff
@@ -482,6 +482,18 @@ def __add_cldap_module(self):
         """
         self.__add_plugin_conf('CLDAP', 'ipa_cldap', 'ipa-cldap-conf.ldif')
 
+    def __allow_trust_attr_mod(self):
+        """
+        aci to allow users to manage own trust attributes
+        """
+        user_aci = paths.UPDATES_DIR + "75-user-trust-attributes.update"
```
Comment (tiran) [inline ipaserver/install/adtrustinstance.py]: I think this is neither necessary nor won't this code work properly.

* ``ipa-server-upgrade`` automatically applies all update files. It should not be necessary to run the update manually.
* update files aren't LDIF files. They look a bit like LDIF, but they have a different syntax. ldapmodify can only process LDIF files. Update files are handled by ``ipaserver.install.ldapupdate.LDAPUpdate``.

File: install/ui/src/freeipa/user.js [hunk 3]
```diff
@@ -1078,6 +1120,7 @@ IPA.user.deleter_dialog = function(spec) {
     return that;
 };
 
+
```
Comment (tiran) [inline install/ui/src/freeipa/user.js]: unrelated change?

File: install/updates/75-user-trust-attributes.update [hunk 4]
```diff
@@ -0,0 +1,4 @@
+# Add an explicit self-service ACI to allow writing to manage trust attributes
+# for the owner of the object
+dn: cn=users,cn=accounts,$SUFFIX
+add:aci:(targetattr = "ipantlogonscript || ipantprofilepath || ipanthomedirectory || ipanthomedirectorydrive")(version 3.0;acl "selfservice:Users can manage their own NT properties";allow (write) userdn = "ldap:///self";)
```
[the remark this answers is not in the document]
Comment (Tiboris) [reply install/updates/75-user-trust-attributes.update]: ok

File: install/ui/src/freeipa/user.js (hunk 3 above)
Comment (Tiboris) [reply install/ui/src/freeipa/user.js]: ok

File: ipaserver/install/adtrustinstance.py (hunk 2 above)
Comment (Tiboris) [reply ipaserver/install/adtrustinstance.py]: Got it.

**Comment by tiran** [comment]
One test is failing: `` test_config.test_command[0000: config_mod: Try to add an unrelated objectclass to ipauserobjectclasses]``

File: API.txt [hunk 5]
```diff
@@ -4882,6 +4882,10 @@ option: Int('gidnumber?')
 option: Str('givenname', cli_name='first')
 option: Str('homedirectory?', cli_name='homedir')
 option: Str('initials?', autofill=True)
+option: Str('ipanthomedirectory?', cli_name='nt_home_directory')
```
Comment (abbra) [inline API.txt]: Please change `cli_name` to `nthomedir`, it will be a pain to use otherwise.

File: API.txt [hunk 6]
```diff
@@ -4882,6 +4882,10 @@ option: Int('gidnumber?')
...
+option: StrEnum('ipanthomedirectorydrive?', cli_name='nt_home_directory_drive', values=[u'A:', u'B:', u'C:', u'D:', u'E:', u'F:', u'G:', u'H:', u'I:', u'J:', u'K:', u'L:', u'M:', u'N:', u'O:', u'P:', u'Q:', u'R:', u'S:', u'T:', u'U:', u'V:', u'W:', u'X:', u'Y:', u'Z:'])
```
Comment (abbra) [inline API.txt]: Please change cli_name to `nthomedrive`

File: ipaserver/plugins/baseuser.py [hunk 7]
```diff
@@ -123,6 +123,24 @@ def fix_addressbook_permission_bindrule(name, template, is_new,
         template['ipapermbindruletype'] = 'anonymous'
 
 
+def update_samba_attrs(ldap, dn, entry_attrs, options):
+    samba_attrs = {'ipantlogonscript', 'ipantprofilepath',
+                   'ipanthomedirectory', 'ipanthomedirectorydrive'}
+
+    for attr in samba_attrs:
+        if options.get(attr, None):
+            if attr in entry_attrs['objectclass']:
```
Comment (abbra) [inline ipaserver/plugins/baseuser.py]: This is wrong -- attribute value cannot be in the `objectclass`. I suspect you wanted to check that `ipaNTUserAttrs` is in the `objectclass`.

File: ipaserver/plugins/baseuser.py [hunk 8]
```diff
@@ -123,6 +123,24 @@ def fix_addressbook_permission_bindrule(name, template, is_new,
...
+                entry = ldap.get_entry(dn)
+                entry[attr] = [options[attr]]
+                ldap.update_entry(entry)
+            else:
+                raise errors.ValidationError(
+                    name=attr,
+                    error=_(
+                        'can not update entry'
```
Comment (abbra) [inline ipaserver/plugins/baseuser.py]: I'd have something more reasonable here: 'Object class ipaNTUserAttrs is missing, entry cannot have Samba attributes.'

File: ipaserver/plugins/baseuser.py [hunk 9]
```diff
@@ -378,6 +397,26 @@ class baseuser(LDAPObject):
...
+        Str('ipantlogonscript?',
+            cli_name='logon_script',
+            label=_('Logon script'),
+            ),
+        Str('ipantprofilepath?',
+            cli_name='profile_path',
+            label=_('Profile path'),
+            ),
+        Str('ipanthomedirectory?',
+            cli_name='nt_home_directory',
```
Comment (abbra) [inline ipaserver/plugins/baseuser.py]: `cli_name='nthomedir'`

File: ipaserver/plugins/baseuser.py [hunk 10]
```diff
@@ -378,6 +397,26 @@ class baseuser(LDAPObject):
...
+            label=_('NT Home Directory'),
+            ),
+        StrEnum('ipanthomedirectorydrive?',
+                cli_name='nt_home_directory_drive',
```
Comment (abbra) [inline ipaserver/plugins/baseuser.py]: `cli_name='nthomedrive'`

File: ipaserver/plugins/baseuser.py (hunk 8 above)
Comment (Tiboris) [reply ipaserver/plugins/baseuser.py]: Yes thats better i will definitelly update it.

File: install/updates/75-user-trust-attributes.update [hunk 11]
```diff
@@ -0,0 +1,4 @@
+# Add an explicit self-service ACI to allow writing to manage trust attributes
+# for the owner of the object
+dn: cn=users,cn=accounts,$SUFFIX
+add:aci:(targetattr = "ipantlogonscript || ipantprofilepath || nthomedir || nthomedrive")(version 3.0;acl "selfservice:Users can manage their own NT properties";allow (write) userdn = "ldap:///self";)
```
Comment (abbra) [inline install/updates/75-user-trust-attributes.update]: the attribute names are different:
```
attributeTypes: (2.16.840.1.113730.3.8.11.8 NAME 'ipaNTHomeDirectory' DESC 'User Home Directory Path' EQUALITY caseIgnoreMatch ORDERING caseIgnoreOrderingMatch SUBSTR caseIgnoreSubstringsMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.15 SINGLE-VALUE X-ORIGIN 'IPA v3' )
attributeTypes: (2.16.840.1.113730.3.8.11.9 NAME 'ipaNTHomeDirectoryDrive' DESC 'User Home Drive Letter' EQUALITY caseIgnoreMatch ORDERING caseIgnoreOrderingMatch SUBSTR caseIgnoreSubstringsMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.15 SINGLE-VALUE X-ORIGIN 'IPA v3' )
```

File: ipaserver/plugins/baseuser.py [hunk 12]
```diff
@@ -123,6 +123,26 @@ def fix_addressbook_permission_bindrule(name, template, is_new,
         template['ipapermbindruletype'] = 'anonymous'
 
 
+def update_samba_attrs(ldap, dn, entry_attrs, **options):
+    samba_attrs = {'ipantlogonscript', 'ipantprofilepath',
+                   'nthomedir', 'nthomedrive'}
```
Comment (abbra) [inline ipaserver/plugins/baseuser.py]: Same here, `nthomedir`, `nthomedrive` are wrong names.

File: ipaserver/plugins/baseuser.py [hunk 13]
```diff
@@ -149,7 +169,8 @@ class baseuser(LDAPObject):
         'ipatokenradiusconfiglink', 'ipatokenradiususername',
         'krbprincipalexpiration', 'usercertificate;binary',
         'krbprincipalname', 'krbcanonicalname',
-        'ipacertmapdata'
+        'ipacertmapdata', 'ipantlogonscript', 'ipantprofilepath',
+        'nthomedir', 'nthomedrive'
```
Comment (abbra) [inline ipaserver/plugins/baseuser.py]: Same here.

File: ipaserver/plugins/baseuser.py [hunk 14]
```diff
@@ -378,6 +399,26 @@ class baseuser(LDAPObject):
...
         ),
+        Str('ipantlogonscript?',
+            cli_name='logon_script',
+            label=_('Logon script'),
+            ),
+        Str('ipantprofilepath?',
+            cli_name='profile_path',
+            label=_('Profile path'),
+            ),
+        Str('nthomedir?',
```
Comment (abbra) [inline ipaserver/plugins/baseuser.py]: Please change the name of the parameter to be in line with the LDAP attribute.

File: ipaserver/plugins/baseuser.py [hunk 15]
```diff
@@ -378,6 +399,26 @@ class baseuser(LDAPObject):
...
+            cli_name='nt_home_directory',
+            label=_('NT Home Directory'),
+            ),
+        StrEnum('nthomedrive?',
```
Comment (abbra) [inline ipaserver/plugins/baseuser.py]: Same here.

File: ipaserver/plugins/internal.py [hunk 16]
```diff
@@ -1483,6 +1483,23 @@ class i18n_messages(Command):
...
+                "title": _(
+                    "User attributes for Active Directory integration"
+                ),
+                "logonscript_tooltip": _(
+                    "Path to a script executed on a Windows system at logon"
+                ),
+                "profilepath_tooltip": _(
+                    "Path to a user profile, in UNC format \\\\server\\share\\"
+                ),
+                "homedir_tooltip": _(
```
Comment (abbra) [inline ipaserver/plugins/internal.py]: .. and here

File: ipaserver/plugins/internal.py [hunk 17]
```diff
@@ -1483,6 +1483,23 @@ class i18n_messages(Command):
...
+                    "Path to a user home directory, in UNC format"
+                ),
+                "homedrive_tooltip": _(
```
Comment (abbra) [inline ipaserver/plugins/internal.py]: .. and here too

File: install/ui/src/freeipa/user.js [hunk 18]
```diff
@@ -361,6 +361,41 @@ return {
...
+                            }
+                        },
+                        {
+                            name: 'ipantprofilepath',
+                            tooltip: {
+                                title: '@i18n:objects.trust_attributes.profilepath_tooltip'
+                            }
+                        },
+                        {
+                            name: 'nthomedir',
```
Comment (abbra) [inline install/ui/src/freeipa/user.js]: Here too

File: install/ui/src/freeipa/user.js [hunk 19]
```diff
@@ -361,6 +361,41 @@ return {
...
+                            tooltip: {
+                                title: '@i18n:objects.trust_attributes.homedir_tooltip'
+                            }
+                        },
+                        {
+                            name: 'nthomedrive',
```
Comment (abbra) [inline install/ui/src/freeipa/user.js]: and here

File: install/ui/src/freeipa/user.js [hunk 20]
```diff
@@ -361,6 +361,41 @@ return {
...
+                            $type: 'select',
+                            options: IPA.create_options([
+                                'A:', 'B:', 'C:', 'D:', 'E:', 'F:', 'G:', 'H:', 'I:',
+                                'J:', 'K:', 'L:', 'M:', 'N:', 'O:', 'P:', 'Q:', 'R:',
+                                'S:', 'T:', 'U:', 'V:', 'W:', 'X:', 'Y:', 'Z:'
+                            ]),
+                            tooltip: {
+                                title: '@i18n:objects.trust_attributes.homedrive_tooltip'
```
Comment (abbra) [inline install/ui/src/freeipa/user.js]: .. and here as well

File: install/ui/src/freeipa/user.js [hunk 21]
```diff
@@ -361,6 +361,41 @@ return {
...
+                        {
+                            name: 'ipantprofilepath',
+                            tooltip: {
+                                title: '@i18n:objects.trust_attributes.profilepath_tooltip'
+                            }
+                        },
+                        {
+                            name: 'nthomedir',
+                            tooltip: {
+                                title: '@i18n:objects.trust_attributes.homedir_tooltip'
```
Comment (abbra) [inline install/ui/src/freeipa/user.js]: here too

File: install/ui/src/freeipa/user.js [hunk 22]
```diff
@@ -361,6 +361,41 @@ return {
...
                         { $type: 'multivalued', name: 'carlicense' }
                     ]
+                },
+                {
+                    name: 'trust attributes',
+                    label: '@i18n:objects.trust_attributes.title',
+                    fields: [{
+                            name: 'ipantlogonscript',
+                            tooltip: {
+                                title: '@i18n:objects.trust_attributes.logonscript_tooltip'
```
Comment (abbra) [inline install/ui/src/freeipa/user.js]: It is better to have these names exactly same as LDAP attributes, so `objects.trust_attributes.ipantlogonscript_tooltip`, to avoid confusion. Same in other places.

**Comment by abbra** [comment]
@Tiboris something I didn't notice in past -- you have to use proper names exactly like their LDAP attributes because this is how mapping is done between the framework and the LDAP storage.

**Comment by Tiboris** [comment]
Closing in favor of #3339.
